tensorflow/contrib/opt/python/training/delay_compensated_gradient_descent.py

```diff
@@ -0,0 +1,254 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""DelayCompensatedGradientDescentOptimizer for TensorFlow."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.python.framework import ops
+from tensorflow.python.ops import state_ops
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import variables
+from tensorflow.python.training import optimizer
+from tensorflow.python.training import training_ops
+
+
+class _RefVariableAsynchronousProcessor(optimizer._RefVariableProcessor):
+    """Processor for Variable."""
+    def update_op_asynchronous(self, optimizer, g, index):
+        if isinstance(g, ops.Tensor):
+            return optimizer._apply_dense(g, self._v, index)
+        else:
+            assert isinstance(g, ops.IndexedSlices), ("Gradient ", g, " is neither a "
+                                                      "tensor nor IndexedSlices.")
+            # pylint: disable=protected-access
+            return optimizer._apply_sparse_duplicate_indices(g, self._v, index)
+
+class _DenseResourceVariableAsynchronousProcessor(optimizer._DenseResourceVariableProcessor):
+    """Processor for dense ResourceVariables."""
```

Comment:
two space indent in all python files

see https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#python-coding-style in general